src/Npgsql/NpgsqlTypes/NpgsqlDateTime.cs

```diff
@@ -1,478 +1,218 @@
-        IComparer<NpgsqlDateTime>, IComparer
+    static class NpgsqlDateTime
     {
-        #region Fields
-
-        readonly NpgsqlDate _date;
-        readonly TimeSpan _time;
-        readonly InternalType _type;
-
-        #endregion
+        internal const int JulianUnixEpochDay = 2440588;
+        internal const int JulianPostgreSqlEpochDay = 2451545;
+
+        internal const int DaysMinValue = 0 - JulianPostgreSqlEpochDay;
+        internal const int DaysMaxValue = 2147483494 - JulianPostgreSqlEpochDay - 1;
+        internal const int DaysNegativeInfinity = int.MinValue;
+        internal const int DaysPositiveInfinity = int.MaxValue;
+
+        internal const long MicrosecondsMinValue = -211813488000000000;
+        internal const long MicrosecondsMaxValue = 9223371331200000000;
+        internal const long MicrosecondsNegativeInfinity = long.MinValue;
+        internal const long MicrosecondsPositiveInfinity = long.MaxValue;
+
+        internal const int MillisecondsPerSecond = 1000;
+        internal const int SecondsPerMinute = 60;
+        internal const int MinutesPerHour = 60;
+        internal const int HoursPerDay = 24;
+
+        internal const int SecondsPerHour = SecondsPerMinute * MinutesPerHour;
+
+        internal const long MicrosecondsPerMicrosecond = 1;
+        internal const long MicrosecondsPerMillisecond = MicrosecondsPerMicrosecond * 1000;
+        internal const long MicrosecondsPerSecond = MicrosecondsPerMillisecond * 1000;
+        internal const long MicrosecondsPerMinute = MicrosecondsPerSecond * SecondsPerMinute;
+        internal const long MicrosecondsPerHour = MicrosecondsPerMinute * MinutesPerHour;
+        internal const long MicrosecondsPerDay = MicrosecondsPerHour * HoursPerDay;
+
+        internal const int DaysPerMonth = 30;
+        internal const int DaysPerYear = 365;
+        internal const int DaysPer4Years = DaysPerYear * 4 + 1;
+        internal const int DaysPer100Years = DaysPer4Years * 25 - 1;
+        internal const int DaysPer400Years = DaysPer100Years * 4 + 1;
+
+        internal const long TicksPerMicrosecond = TimeSpan.TicksPerMillisecond / MicrosecondsPerMillisecond;
+        internal const long TicksPerDay = TicksPerMicrosecond * MicrosecondsPerDay;
+
+        // Offset is calculated as a difference in microseconds between
+        // 0000-01-01 and 2000-01-01 multiplied by ticks per microsecond.
+        internal const long TicksOffset = 63082281600000000 * TicksPerMicrosecond;
+
+        internal static int GetDatePart(long microseconds) =>
+            microseconds switch
+            {
+                MicrosecondsPositiveInfinity => DaysPositiveInfinity,
+                MicrosecondsNegativeInfinity => DaysNegativeInfinity,
+                _ => (int)(microseconds / MicrosecondsPerDay)
+            };
 
-        #region Constants
+        internal static long GetTimePart(long microseconds) =>
```

Comment:
All this methods do make `NpgsqlDateTime` look like some kind of helper class (which it's definitely not). Maybe we should move them somewhere else?

Reply:
It's exactly a helper class providing most of the functions and constants to data types outside of it.

Reply:
I think @vonzshik means that it's weird for this to be both the user-facing NpgsqlDateTime, and a helper class with lots of static methods used by other types. I agree it would be good to move the latter out to an internal DateTimeUtils (or whatever), to keep NpgsqlDateTime about being a user-facing timestamp type.

Reply:
Forget about that nasty named types we had before, there's `NpgsqlTimestamp` which represents `timestamp` as is and has the right and corresponding name. `NpgsqlDateTime` is an internal helper type or should be so if rebase wasn't broke it.

Reply:
OK, makes sense to me. In that case it's probably better to just rename it to DateTimeHelpers or similar.

Reply:
@roji yeah, thanks for the translation)

Agree with renaming it.

Reply:
It breaks file ordering, so not happy with renaming, but if we are going to remove `Npgsql` from all types here, than will do it. All types here have to specific name to collide with something else outside of the driver.

Reply:
> but if we are going to remove Npgsql from all types here, than will do it

Unless I'm mistaken, the helpers class would be completely internal, so a name like DateTimeHelpers is appropriate. However, we definitely cannot remove Npgsql from all types (I don't think Npgsql can expose types named "Date", "Timestamp" etc.).

> It breaks file ordering

?